keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1066 @@
+of the `LoadBalancerServing` condition. (For example, if the user
+changes the service's `LoadBalancerSourceRanges`, and the cloud
+provider updates the load balancer for this, it must update the
+`LoadBalancerServing` condition's `LastTransitionTime`, even if the
+value of `LoadBalancerServing` did not toggle from `True` to `False`
+and back to `True`.)
+
+The well-known `Reason` values for the `LoadBalancerServing` condition
+are:
+
+  - `Status=True, Reason=Serving`
+
+  - `Status=False, Reason=Provisioning`: The load balancer is not
+    currently serving because it is being provisioned /
+    re-provisioned.
+
+  - `Status=False, Reason=Unsupported`: No load balancer could be
+    provisioned, because the cloud provider doesn't support some
+    feature of this Service (explained further in `Message`).
+
+  - `Status=False, Reason=Infrastructure`: No load balancer could be
+    provisioned, because of a problem with the underlying cloud
+    infrastructure (explained further in `Message`).
+
+#### The `LoadBalancerProvisioning` Condition
+
+The `LoadBalancerProvisioning` condition, when `True`, indicates that
+the cloud provider is currently performing a time-consuming
+resource-provisioning step for the load balancer. When
+`LoadBalancerProvisioning` is `True`, the load balancer is either not
+handling connections, or else is still handling connections according
+to the semantics of a previous state of the Service (with the value of
+the `LoadBalancerServing` condition distinguishing these two cases).
+
+The cloud provider should set a `LoadBalancerProvisioning` condition
+immediately upon first becoming aware of a `Type: LoadBalancer`
+Service. At all times, the condition should be `True` if the cloud
+provider is waiting for the completion of an operation to bring the
+load balancer's behavior in line with the current state of the Service
+object, and `False` if it not. (There are no defined semantics for an
+`Unknown` value of this condition.)
+
+Any time any load-balancer-related field (or provider-specific
+annotation) of the Service object changes, the cloud provider must
+update at least the `LastTransitionTime` of the
+`LoadBalancerProvisioning` condition, to indicate that it has seen the
+update, even if the value of `LoadBalancerProvisioning` remains
+`False`.
+
+The well-known `Reason` values for the `LoadBalancerProvisioning`
+condition are:
+
+  - `Status=True, Reason=Provisioning`
+
+  - `Status=False, Reason=Complete`: The cloud provider has finished
+    trying to provision the load balancer (whether successfully or
+    unsuccessfully, as indicated by the `Message` and by the
+    `LoadBalancerServing` condition).
+
+#### Service Lifecycle and Interaction of the Provisioning/Serving Conditions
+
+When a load balancer Service is first created (or a non-load-balancer
+Service is converted to `Type: LoadBalancer`), one of three things
+should immediately happen:
```

Comment:
how immediately?  If it takes 10 seconds for the controller to act, is that OK?

Reply:
Why would the controller take 10 seconds to act?

The intent is "the controller should just do this before it starts doing anything potentially-time-consuming".

In the e2e tests we currently assume that if you make a change to a pod or service, that the endpointslice controller will act on that change "immediately" and the test will be able to observe a change to the corresponding EndpointSlice object within 30 seconds. I was planning to use the same requirement for cloud-providers setting the intitial conditions.

Reply:
The loadbalancer controller interacts with the cloud API and infra, this calls to the cloud APIs use to be sync and block the workers ... on very large clusters this can take ~ 1 hour per LoadBalancer.
In our scale tests for loadbalancers we have to define at least the same number of workers in the loadbalancer controller as the number of loadbalancers that are going to be created in parallel on the tests, otherwise, you are not able to process any event until some worker is available.

Reply:
This KEP would (necessarily) make that architecture "illegal". One of the major points of the conditions is to be able to distinguish between "no cloud provider is running" and "the cloud provider is taking an hour to allocate your load balancer" (and "the cloud provider already determined that it's not going to be able to allocate your load balancer for some reason"). So the cloud provider needs an informer event loop that doesn't block on the cloud API.

Reply:
there is no need to change any architecture at all, is just the conditions need to be updated before it gets to the worker ...